### PR TITLE
Subclasses/client proxies - skip delegation if not constructed yet

### LINKED ISSUE #2774: Client proxies - skip delegation if proxy is not constructed yet
The proxy should not delegate if the proxy constructor has not finished yet. See Weld implementation: https://github.com/weld/core/blob/master/impl/src/main/java/org/jboss/weld/bean/proxy/ProxyFactory.java#L686-L714.

### PR DESCRIPTION
**Pull request by mkouba**
- resolves #2774

### CODE REVIEW

File: independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/interceptors/constructor/SubclassConstructorGuardTest.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2018 Red Hat, Inc.
```
Comment (cescoffier) [inline independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/interceptors/constructor/SubclassConstructorGuardTest.java]: 2019?

**Review by cescoffier — COMMENTED** [review]
The license header year must be updated.

**Comment by mkouba** [comment]
I'm going to update the licence header but the CI fails when testing native images. I think I will need to handle the case where a client proxy does not extend an impl class but an interface or an abstract class.

**Review by manovotn — COMMENTED** [review]
Solution looks good, but shouldn't we add a test for producers so that we needn't wait for native build failures to discover that?
If you want me to, I can add some.

**Comment by mkouba** [comment]
I think that a test for producers would not reveal this particular issue. However, it seems the `hibernate-orm-panache` native image tests are broken too. So I'll need to investigate further.

**Comment by mkouba** [comment]
So it seems there is a problem with interface default methods (again). But I'm not sure what the real problem is... yet.

**Comment by mkouba** [comment]
Hm, for `PanacheRepositoryBase` implementations some methods (such as `streamAll()`) are overrided/implemented using the bytecode transformer. As a result, the default method declared on `PanacheRepositoryBase` is probably removed during DCE and since we don't know that it's overriden we delegate to the original default method. Nevertheless, I don't know why `native-image` does not consider the client proxy when resolving the elements. In other words, I think it should not remove the method if it's used in a client proxy class. It could be a `native-image` bug. I'll continue with investigation tomorrow.

**Comment by mkouba** [comment]
It's actually a bug in gizmo: https://github.com/quarkusio/gizmo/issues/9

I'm going to leave out the interface default methods support for now.

**Comment by mkouba** [comment]
So the TCK CI build fails with:
> [ERROR]   AsyncMethodTest>Arquillian.run:138->testAsyncInvocationInterceptorProvider:234 expected [21] but found [null]
(Quarkus - TCK - MicroProfile REST Client)

Which is really weird because this branch does not contain this TCK runner. @cescoffier @geoand any ideas?

**Comment by geoand** [comment]
Unfortunately no... However there are various things wrong with the REST Client TCK so I would suggest we exclude that test for the time being and then later on spend some serious time bringing it in line with the TCK

**Comment by mkouba** [comment]
The thing is I cannot exclude the test in my branch - I have no rest client runner there ;-). Of course, I could rebase and exclude but I wonder if it would make sense to understand what azure pipelines actually do.

**Comment by geoand** [comment]
Sorry, got sidetracked... I see you go past the problem :)

**Comment by mkouba** [comment]
Yep, looks like some intermittent failure. We should probably investigate/exclude the test anyway.

@cescoffier Do we merge the PR into master first and run the build there? I thought we run the tck in the PR branch but obviously I was wrong.

**Review by manovotn — APPROVED** [review]
Looks good now!

**Pull request merged**